beamer/chain.py

```diff
@@ -84,7 +84,10 @@ def _thread_func(self) -> None:
         )
 
         fetcher = EventFetcher(self._web3, self._contracts, self._deployment_block)
-        events = fetcher.fetch()
+        current_block = self._web3.eth.block_number
+        events = []
+        while fetcher.synced_block < current_block:
+            events = fetcher.fetch()
```

Comment:
I think this could lose events in case the first `fetch` call returns a nonempty list, but still did not manage to fetch until `current_block`, due to a `ConnectionError`. Since we overwrite `events` on each iteration, we would lose all but the last batch of events.

However, I'm thinking moving this 3-liner to `EventFetcher` would be more appropriate since
`EventFetcher.fetch` semantics has been to fetch until the current block. In that case you would not even
need the `synced` property and the current code in monitor could stay the same.

Something like:
```python
    def fetch(self) -> list[Event]:
           events = []
           while fetcher.synced_block < current_block:
               events.extend(self._fetch())
```
(here `self._fetch` would be the current `EventFetcher.fetch` -- yeah, naming is not perfect...)

What do you think?

Additional concern here is that there's a possibility that we end up with multiple `LatestBlockUpdatedEvent` , instead of just one, ideally. But that would be trivial to filter out in even fetcher and we could just leave the last 
(most recent one) `LatestBlockUpdatedEvent` in the resulting list.

Reply:
Moving this into the EventFetcher would result in a change in the logic that a connection error does not break and return an empty array. It would get stuck in the loop until it was fetched. I actually liked the approach that during normal runtime we just return on a ConnectionError and wait for the next cycle. 
The different desired behavior is only on sync

Reply:
> I think this could lose events in case the first fetch call returns a nonempty list, but still did not manage to fetch until current_block, due to a ConnectionError. Since we overwrite events on each iteration, we would lose all but the last batch of events.

🤦

Reply:
> Additional concern here is that there's a possibility that we end up with multiple LatestBlockUpdatedEvent , instead of just one, ideally. But that would be trivial to filter out in even fetcher and we could just leave the last
(most recent one) LatestBlockUpdatedEvent in the resulting list.

I don't see a problem with that. The behavior on processing LatestBlockUpdatedEvent is exactly equal to if it came in different cycles. A more recent event overwrites the older data.
Since it comes in the correct order so there should be no problem.

Reply:
> I don't see a problem with that. The behavior on processing LatestBlockUpdatedEvent is exactly equal to if it came in different cycles. A more recent event overwrites the older data.
> Since it comes in the correct order so there should be no problem.

Right, I was mostly worried about being possibly less efficient in case of multiple events, but I guess as things stand, it's not a problem at all.